### ISSUE TITLE
Allow custom dialer function or custom net.Conn on client creation

### ISSUE THREAD

**Issue by jonathanpittman** [vmware/govmomi#1305]
I hope I just missed something really simple.

I am working with an environment that involves SSHing to a specific server in order to access a network domain that contains a vcenter server and esx hosts.  I can do port forwarding on the ssh connection to the vcenter server or an esx host and point govc at the locally forwarded port.  This works, but I have to set insecure to true.  I would like to be able to do this without using insecure mode.  And I want to be able to use the host's address on that remote network which is not resolvable on my local network.

I am making my way through the govmomi API to write some of my own tooling for tasks.  In doing so, I am using the [crypto/ssh](https://golang.org/x/crypto/ssh) package to do ssh to the initial server.  From there, within the same running binary, I can use an [ssh.Client's Dial method](https://golang.org/x/crypto/ssh#Client.Dial) to open network connections to hosts on the remote network (i.e. the remote half of an ssh tunnel) and get a net.Conn back.  The ssh.Client's Dial method has the same signature as [net.Dialer's Dial](https://golang.org/pkg/net/#Dialer.Dial). This is also the same as [http.Transport's Dial field](https://golang.org/pkg/net/http/#Transport.Dial).

`func(network, addr string) (net.Conn, error)`

I tried doing something like this...

```
client, err := govmomi.NewClient(ctx, addr, true)
if err != nil {
	return nil, err
}

// (*govmomi.Client).(*vim25.Client).(*soap.Client).(http.Client).(http.RoundTripper).(*http.Transport)
if t, ok := client.Client.Client.Client.Transport.(*http.Transport); ok && dialFn != nil {
	t.Dial = dialFn // e.g. sshClient.Dial
	// or t.Dial = func(network, addr string) (net.Conn, error) { return sshTunnelConn, nil}
}
err := client.Login(ctx, userCreds)
```

In one case, I tried to pass in the ssh Client's Dial method as the Dial function.  In another case, I used the ssh Client to Dial a tcp connection to the remote vcenter instance (successfully), and then make a custom Dial function that returns that conn each time.

However, I think that tls.Dial is being called early on during the call to govmomi.NewClient.  And when this happens, it is unable to resolve the address, because [tls.Dial](https://golang.org/pkg/crypto/tls/#Dial) is trying to resolve the address locally.  And of course the vcenter's host address is only resolvable on the remote network, not my local network. So, it produces an error before I ever get to reassign the Dial function.

I would like the code to be able to use the ssh Client's Dial function to handle the resolution and dialing of the remote host (pre-tls and pre-auth). This will let me connect to separate instances of vcenter (e.g. dev & prod) by just passing in the different host name each time. If I have to dial the connection separately and then pass in the resulting net.Conn for tls and auth, that is also fine.  Something like this might do the trick.

```
tlsConn := tls.Client(sshTunnelConn, &tlsCfg)
tlsConn.Handshake()
```

However, the inner parts of dialing (in govmomi, vim25, and soap) are tucked away enough that I am not sure I can easily swap them out without changes to the code.

**Comment by dougm** [comment]
`govmomi.Client` just adds a few shortcuts, you can use `soap.Client` and `vim25.Client` directly in this case.  For example:
```go
		soapClient := soap.NewClient(u, false)

		soapClient.Transport.(*http.Transport).DialContext = dialFn

		vimClient, err := vim25.NewClient(ctx, soapClient)
		if err != nil {
			log.Fatal(err)
		}

		err = session.NewManager(vimClient).Login(ctx, u.User)
		if err != nil {
			log.Fatal(err)
		}
```

**Comment by jonathanpittman** [comment]
Thanks for leading me in the right direction.

This still results in
```
Post https://vsphere:443/sdk: dial tcp: lookup vsphere: no such host
```
after calling vim25.NewClient(ctx, soapClient).

`vim25.NewClient`
calls `methods.GetServiceContent(ctx, rt)`
calls `methods.RetrieveServiceContent(ctx, r, &req)`
calls `r.RoundTrip(ctx, &reqBody, &resBody)` (where r is a soap.RoundTripper fulfilled by soap.Client.RoundTrip)
calls `soap.Client.Do`
calls `c.Client.Do(req.WithContext(ctx))` (where c is a *soap.Client)

And that c.Client is an http.Client that will eventually call http.Client.Transport.DialTLS because it is set to soap.Client.dialTLS here `https://github.com/vmware/govmomi/blob/master/vim25/soap/client.go#L145`.

I got this working by also doing this `soapClient.Transport.(*http.Transport).DialTLS = nil`. According to the http package:

```If DialTLS is nil, Dial and TLSClientConfig are used.```

It appears that DialContext is actually used if it is present.

I now get `x509: certificate signed by unknown authority`, but that is going in the right direction at least.  I am guessing I can do soapClient.SetRootCAs or something along those lines. In the mean time, setting insecure to true lets this go through.

However, I am wondering if I should be recreating soap.Client.dialTLS to do custom dialing instead. Is there anything special about soap.Client.dialTLS that I need to replicate (e.g. the Thumprint bits)?  Or can it be ignored as long as the certs and hostname are all valid?

So, to sum it up, this works:
```
	soapClient := soap.NewClient(addrURL, true)
	soapClient.Transport.(*http.Transport).DialTLS = nil
	soapClient.Transport.(*http.Transport).Dial = dialFn
	soapClient.Transport.(*http.Transport).DialContext = func(ctx context.Context, network, address string) (net.Conn, error) {
		// handle context
		return dialFn(network, address)
	}
	vimClient, err := vim25.NewClient(ctx, soapClient)
	if err != nil {
		return nil, err
	}
	client := &govmomi.Client{
		Client:         vimClient,
		SessionManager: session.NewManager(vimClient),
	}

	return client, client.Login(ctx, userCreds)
```
And once I add the right RootCAs to the tls config, I should be able to set insecure to false.

But should I recreate soap.Client.dialTLS to do custom dialing?
